### PR TITLE
Update minion task metadata ZNode path

### PR DESCRIPTION
**Pull request by zhtaoxiang**
## What does this PR do?

This PR updates minion task metadata ZNode path from `/MINION_TASK_METADATA/${taskType}/${tableNameWithType}` to `/MINION_TASK_METADATA/${tableNameWithType}/${taskType}`.

The logic is backward compatible: if there exists minion task metadata in ZNode path `/MINION_TASK_METADATA/${tableNameWithType}/${taskType}`, it keeps using it; otherwise, it uses the new ZNode path.

## How was the code tested?
1. It was tested by unit tests
2. It was tested in a local Pinot cluster: in the case that (1) there exists minion task metadata in ZNode path `/MINION_TASK_METADATA/${taskType}/${tableNameWithType}`, and (2) there does not exists minion task metadata.

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java [hunk 1]
```diff
@@ -130,7 +130,13 @@ public static String constructPropertyStorePathForSegmentLineage(String tableNam
     return StringUtil.join("/", PROPERTYSTORE_SEGMENT_LINEAGE, tableNameWithType);
   }
 
-  public static String constructPropertyStorePathForMinionTaskMetadata(String taskType, String tableNameWithType) {
+  public static String constructPropertyStorePathForMinionTaskMetadataNewFormat(
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java:133]: Suggest making this `constructPropertyStorePathForMinionTaskMetadata` and changing the current one to `constructPropertyStorePathForMinionTaskMetadataDeprecated` and annotate it as deprecated

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java:133]: Please check the code style setting. Seems it is not using the Pinot Style

File: pinot-common/src/main/java/org/apache/pinot/common/minion/MinionTaskMetadataUtils.java [hunk 2]
```diff
@@ -36,12 +36,25 @@ private MinionTaskMetadataUtils() {
   }
 
   /**
-   * Fetches the ZNRecord for the given minion task and tableName, from MINION_TASK_METADATA/taskName/tableNameWthType
+   * Fetches the ZNRecord for the given minion task and tableName. Fetch from the old path
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/MinionTaskMetadataUtils.java:39]: I think we should try the new path first, then fall back to the old one if the new one does not exist?

File: pinot-common/src/main/java/org/apache/pinot/common/minion/MinionTaskMetadataUtils.java [hunk 3]
```diff
@@ -51,27 +64,46 @@ public static ZNRecord fetchTaskMetadata(HelixPropertyStore<ZNRecord> propertySt
...
    */
   public static void deleteTaskMetadata(HelixPropertyStore<ZNRecord> propertyStore, String taskType,
       String tableNameWithType) {
-    String path = ZKMetadataProvider.constructPropertyStorePathForMinionTaskMetadata(taskType, tableNameWithType);
-    if (!propertyStore.remove(path, AccessOption.PERSISTENT)) {
+    String newPath = ZKMetadataProvider.constructPropertyStorePathForMinionTaskMetadataNewFormat(
+        taskType, tableNameWithType);
+    String oldPath = ZKMetadataProvider.constructPropertyStorePathForMinionTaskMetadata(
+        taskType, tableNameWithType);
+    if (!propertyStore.remove(newPath, AccessOption.PERSISTENT)
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/MinionTaskMetadataUtils.java:77]: We want to remove from both if both of them exist. Current implementation will short-circuit the second part

**Review by Jackie-Jiang — COMMENTED** [review]
Ideally, we want to automatically move the ZNRecord from the old path to the new path. When reading the ZK metadata, if we find the old path exist:
1. If the new one also exist, remove the old one
2. If the new one does not exist, copy the old record to the new path, then remove the old one

File: pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java (hunk 1 above)
Comment (zhtaoxiang) [reply pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java:133]: interesting, the format does not take effects until I manually reformat the file 🤔

File: pinot-common/src/main/java/org/apache/pinot/common/minion/MinionTaskMetadataUtils.java (hunk 3 above)
Comment (zhtaoxiang) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/MinionTaskMetadataUtils.java:77]: good catch!

File: pinot-common/src/main/java/org/apache/pinot/common/minion/MinionTaskMetadataUtils.java (hunk 2 above)
Comment (zhtaoxiang) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/MinionTaskMetadataUtils.java]: okay, updated!

**Comment by zhtaoxiang** [comment]
> Ideally, we want to automatically move the ZNRecord from the old path to the new path. When reading the ZK metadata, if we find the old path exist:
> 
> 1. If the new one also exist, remove the old one
> 2. If the new one does not exist, copy the old record to the new path, then remove the old one

We can do 1 safely. 
However, 2 is not easy, since the copy-remove operation needs to be atomic. Otherwise, we cannot guarantee that the old and the new ZNodes hold the same data, so we will never know when it's safe to remove the old ZKNode.

I'd prefer we keep the current logic and discuss how to implement the proposal correctly.

File: pinot-common/src/main/java/org/apache/pinot/common/minion/RealtimeToOfflineSegmentsTaskMetadata.java [hunk 4]
```diff
@@ -26,7 +26,7 @@
  * The <code>watermarkMs</code> denotes the time (exclusive) upto which tasks have been executed.
  *
  * This gets serialized and stored in zookeeper under the path
- * MINION_TASK_METADATA/RealtimeToOfflineSegmentsTask/tableNameWithType
+ * MINION_TASK_METADATA/tableNameWithType/RealtimeToOfflineSegmentsTask
```
Comment (jackjlli) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/RealtimeToOfflineSegmentsTaskMetadata.java:29]: `MINION_TASK_METADATA/t${ableNameWithType}/RealtimeToOfflineSegmentsTask`

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/realtimetoofflinesegments/RealtimeToOfflineSegmentsTaskGenerator.java [hunk 5]
```diff
@@ -57,7 +57,7 @@
  *
  * Steps:
  *  - The watermarkMs is read from the {@link RealtimeToOfflineSegmentsTaskMetadata} ZNode
- *  found at MINION_TASK_METADATA/RealtimeToOfflineSegmentsTask/tableNameWithType
+ *  found at MINION_TASK_METADATA/tableNameWithType/RealtimeToOfflineSegmentsTask
```
Comment (jackjlli) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/realtimetoofflinesegments/RealtimeToOfflineSegmentsTaskGenerator.java:60]: Make it consistent as the other comments.

File: pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java [hunk 6]
```diff
@@ -131,6 +130,12 @@ public static String constructPropertyStorePathForSegmentLineage(String tableNam
   }
 
   public static String constructPropertyStorePathForMinionTaskMetadata(String taskType, String tableNameWithType) {
```
Comment (npawar) [inline pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java:132]: nit: would suggest to keep the order of the params same as the order in the path for this new one (tableNameWithType, taskType)

**Review by npawar — APPROVED** [review]
minor comment. lgtm otherwise

**Pull request merged**